x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardIndexer.java

```diff
@@ -186,20 +198,44 @@ private void checkCancelled() {
                 task.getNumIndexed(),
                 task.getNumFailed()
             );
+            task.setRollupShardIndexerStatus(RollupShardIndexerStatus.CANCELLED);
             throw new TaskCancelledException(format("Shard %s rollup cancelled", indexShard.shardId()));
         }
     }
 
     private BulkProcessor2 createBulkProcessor() {
         final BulkProcessor2.Listener listener = new BulkProcessor2.Listener() {
+            private long bulkStartTime;
+
             @Override
             public void beforeBulk(long executionId, BulkRequest request) {
                 task.addNumSent(request.numberOfActions());
+                task.setBeforeBulkInfo(
+                    new RollupBeforeBulkInfo(
+                        System.currentTimeMillis(),
+                        executionId,
+                        request.estimatedSizeInBytes(),
+                        request.numberOfActions()
+                    )
+                );
+                bulkStartTime = System.currentTimeMillis();
             }
 
             @Override
             public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+                long bulkDurationInMillis = System.currentTimeMillis() - bulkStartTime;
                 task.addNumIndexed(request.numberOfActions());
+                task.setAfterBulkInfo(
+                    new RollupAfterBulkInfo(
+                        System.currentTimeMillis(),
+                        executionId,
+                        bulkDurationInMillis,
+                        response.getIngestTookInMillis(),
```

Comment:
This only keeps track of the latest details, which is useful. 
But maybe also keep track of mean took time / bulk duration? (using `MeanMetric` class)

Reply:
To do this I need to keep a running count/sum class for bulk requests instead of having just the latest one.